Enabled commutative property for MatrixElement

arihantparsoya opened the pull request:
This is fix for #11077 

Previously the `is_commutative` property was `None` and hence Sympy wasnt able to simplify the determinant expression.
- After setting the `is_commutative` property to `True`, the determinant expression is being simplified.
- None of the tests fail after changing the `is_commutative` property

arihantparsoya committed:
Enabled commutative property for MatrixElement

parsoyaarihant commented:
@asmeurer , can you review this?

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master' into patch-for-#11062

asmeurer commented:
Looks like some tests failed. Those should be investigated.

asmeurer commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -338,7 +338,7 @@ class MatrixElement(Expr):
     j = property(lambda self: self.args[2])
     _diff_wrt = True
-
+    is_commutative = True
```
Keep a blank line before function definitions.

asmeurer commented:
You should add a test for the original issue.

arihantparsoya committed:
typo

arihantparsoya committed:
Changed output of test cases

Printing tests which give the correct output but different in string
values are modified:

example:
M[6] = 2*q[4]*1/q[1]; is equivalent to
M[6] = 2*q[4]/q[1];

parsoyaarihant commented:
There are in total 4 test failures, one is related to the PR [#9364](https://github.com/sympy/sympy/pull/9364) and the remaining 3 are related to printing which I fixed in my last commit. I will look into [#9364](https://github.com/sympy/sympy/pull/9364) and fix the last remining test case.

jksuom commented:
I wonder if there could be a way of making matrix elements commutative on demand. Matrices with elements in non-commutative rings are generally quite acceptable. Commutativity is needed only in some special cases such as for determinants.

arihantparsoya committed:
Made correction in the test

As the MatrixElement is_commutative property is set to true, the
expression in line 2283 wasn’t being interpreted correctly and the
cancel function was returning unsimplified output previously

arihantparsoya committed:
typo error

parsoyaarihant commented:
@asmeurer , should I add the tests for the determinant of the matrix (i.e the issue for determinant which u originally posted) or just the commutability of the MatrixElement?

arihantparsoya committed:
Added tests

Added test for commutative property of MatrixElement

parsoyaarihant commented:
@asmeurer , can you review this?

asmeurer commented on sympy/polys/tests/test_polytools.py:
```diff
@@ -2881,7 +2882,7 @@ def test_cancel():
     assert cancel(M[0,0] + 7) == M[0,0] + 7
     expr = sin(M[1, 4] + M[2, 1] * 5 * M[4, 0]) - 5 * M[1, 2] / z
-    assert cancel(expr) == expr
+    assert simplify(cancel(expr)) == expr
```
This test seems wrong. The point of the function is to test `cancel`. What does this do without `simplify` here?

asmeurer replied:
I guess it should return `(z*sin(M[1, 4] + M[2, 1] * 5 * M[4, 0]) - 5 * M[1, 2]) / z`. If that's what it's giving, we should test that.

parsoyaarihant replied:
yeah it returns `cancel(expr) = (z*sin(M[1, 4] + M[2, 1] * 5 * M[4, 0]) - 5 * M[1, 2]) / z` and previously it was returning the expression `sin(M[1, 4] + 5*M[2, 1]*M[4, 0]) - 5*M[1, 2]/z`, please note that the whole numerator is encloses in a bracket after `is_commutative` is set to `True` hence I used `simplify`

asmeurer replied:
`(z*sin(M[1, 4] + M[2, 1] * 5 * M[4, 0]) - 5 * M[1, 2]) / z` is the right answer for cancel to return. The test should be

```py
assert cancel(expr) == (z*sin(M[1, 4] + M[2, 1] * 5 * M[4, 0]) - 5 * M[1, 2]) / z
```

Calling `simplify` in the function that tests the output of `cancel` defeats the purpose.

asmeurer commented:
I would add a test for the commutativity (which you already have), and a test for the determinant, to make sure that that works.

arihantparsoya committed:
added test for determinant of matrix

parsoyaarihant commented:
@asmeurer , I added the test for determinant of the matrix.

asmeurer commented:
Just the comment on the `cancel` test. The rest looks good.

arihantparsoya committed:
fixed test_polytools.py tests

parsoyaarihant commented:
@asmeurer , I fixed the test.

parsoyaarihant commented:
@asmeurer , can you review this again?

asmeurer commented:
Looks good, thanks.

asmeurer commented:
Oh, one thing. In the future, if you fix an issue, put "Fixes #11077" in one of your commit messages. That way, when the PR is merged, the issue will be automatically closed.
